### PR TITLE
Implement area calculation for footprint of 3D objects

### LINKED ISSUE #333: bug in calculate_area
The function `calculate_area` takes `features, mask` as argument. If mask is 3D (i.e. incl. time), `calculate_area` fails because internally calculates an area field (`np.array`) which is 2D. Both `mask` and `area` are provided to the`scipy.ndimage` function `labeled_comprehension` which however assumes that the 1st two arguments have the same shape. confict -> 2D != 3D

Here is part of the error
```python
File [~/TROPOS/proj/2022-06-tobac-devel/tobac/tobac/analysis.py:960](https://file+.vscode-resource.vscode-cdn.net/home/senf/TROPOS/proj/2022-06-tobac-devel/tobac/examples/Basics/~/TROPOS/proj/2022-06-tobac-devel/tobac/tobac/analysis.py:960), in calculate_area(features, mask, method_area)
    957 else:
    958     raise ValueError("method undefined")
--> 960 feature_areas = labeled_comprehension(
    961     area, mask.data, features["feature"], np.sum, area.dtype, np.nan
    962 )
    964 features["area"] = feature_areas
    966 return features

File [~/TROPOS/proj/2022-06-tobac-devel/tobac-env-py38/lib/python3.8/site-packages/scipy/ndimage/_measurements.py:511](https://file+.vscode-resource.vscode-cdn.net/home/senf/TROPOS/proj/2022-06-tobac-devel/tobac/examples/Basics/~/TROPOS/proj/2022-06-tobac-devel/tobac-env-py38/lib/python3.8/site-packages/scipy/ndimage/_measurements.py:511), in labeled_comprehension(input, labels, index, func, out_dtype, default, pass_positions)
    509     input, labels = numpy.broadcast_arrays(input, labels)
    510 except ValueError as e:
--> 511     raise ValueError("input and labels must have the same shape "
    512                         "(excepting dimensions with width 1)") from e
    514 if index is None:
    515     if not pass_positions:

ValueError: input and labels must have the same shape (excepting dimensions with width 1)
```

A solution would be to care in the calculations that `area`has always the same shape as `mask`.

As `calculate_area` is not included in the testing I guess it is not really used by the group, right?

### PR DESCRIPTION
**Pull request by w-k-jones**
Resolves #333 

Changes `calculate_area` to use bulk statistics for area calculation, and enables area calculation for both 2D and 3D tracked objects. Area is now not calculated with a time dimension when using lat/lon coords, reducing memory usage. A new `collapse_dim` keyword has been added to `get_statistics_from_mask`, allowing one or more dimensions of the input labels to be collapsed. This, for example, allows calculation of the 2D area of the footprint of a 3D tracked object. 

I have also reorganised `analysis` into a package, with a separate file for spatial functions. Functions other than `calculate_area` are identical except for updates to imports

* [x] Have you followed our guidelines in CONTRIBUTING.md? 
* [x] Have you self-reviewed your code and corrected any misspellings? 
* [x] Have you written documentation that is easy to understand?
* [x] Have you written descriptive commit messages? 
* [x] Have you added NumPy docstrings for newly added functions? 
* [x] Have you formatted your code using black? 
* [x] If you have introduced a new functionality, have you added adequate unit tests?
* [x] Have all tests passed in your local clone? 
* [x] If you have introduced a new functionality, have you added an example notebook?
* [x] Have you kept your pull request small and limited so that it is easy to review? 
* [x] Have the newest changes from this branch been merged?

### CODE REVIEW

File: tobac/analysis/analysis.py
Comment (fziegner) [inline tobac/analysis/analysis.py:937]: There still seems to be an issue when using `method_area="xy"` for the area calculation. I tested it with the segmentation results of the OLR_Tracking_model notebook and got this error:

Screenshot 2024-01-31 143227

Switching the lines in `area = np.outer(...)` fixes it on my end.

Comment (w-k-jones) [reply tobac/analysis/analysis.py:937]: Thanks for catching this, have fixed it now

File: tobac/__init__.py
Comment (fsenf) [reply tobac/__init__.py:44]: @w-k-jones : Nice work!

Two suggestions:
- I don't really like analysis.analysis . It should be more specific. What about renaming it to analysis/statistics.py?
- I would recommend to `analysis/__init__.py` the following lines:
```python
from .analysis import *
from .spatial import *
```
of after rneaming
```python
from .statistics import *
from .spatial import *
```
- it is already the case that refactoring has broken things in the notebook `tobac/examples/Basics/Idealized-Case-1_Tracking-of-a-Test-Blob-in-2D.ipynb` --> Better the keep the old structure for a while.

**Comment by fsenf** [comment]
@w-k-jones : I updated the PR and made some improvements to notebook ` tobac/examples/Basics/Idealized-Case-1_Tracking-of-a-Test-Blob-in-2D.ipynb` in your fork: https://github.com/w-k-jones/tobac/pull/1

**Comment by fsenf** [comment]
I tried to understand how collapse_dim is working in get_statistics. ... well coded, but super hard to digest by just looking at the code. Therefore, I mainly looked at the existing tests and played around with the shapes. 

A horseshoe shape gives me unexpected results:

```python
    test_labels = np.array(
        [
            [0, 0, 0, 0, 0],
            [0, 1, 1, 1, 0],
            [0, 1, 0, 1, 0],
            [0, 1, 2, 1, 0],
            [0, 0, 0, 0, 0],
        ],
        dtype=int,
    )

    test_values = np.array([0.3, 0.5, 0.75, 1, 1.5])

    test_features = pd.DataFrame(
        {
            "feature": [1, 2],
            "frame": [0, 0],
            "time": [
                datetime(2000, 1, 1),
                datetime(2000, 1, 1),
            ],
        }
    )
    statistics_sum = {"sum": np.sum}

    expected_sum_result_axis1 = np.array([2.25, 0.75])

    output_collapse_axis1 = tb_utils.get_statistics(
        test_features,
        test_labels,
        test_values,
        statistic=statistics_sum,
        collapse_axis=1,
    )

    assert np.all(output_collapse_axis1["sum"] == expected_sum_result_axis1)


---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[27], [line 36](vscode-notebook-cell:?execution_count=27&line=36)
     [26](vscode-notebook-cell:?execution_count=27&line=26) expected_sum_result_axis1 = np.array([2.25, 0.75])
     [28](vscode-notebook-cell:?execution_count=27&line=28) output_collapse_axis1 = tb_utils.get_statistics(
     [29](vscode-notebook-cell:?execution_count=27&line=29)     test_features,
     [30](vscode-notebook-cell:?execution_count=27&line=30)     test_labels,
   (...)
     [33](vscode-notebook-cell:?execution_count=27&line=33)     collapse_axis=1,
     [34](vscode-notebook-cell:?execution_count=27&line=34) )
---> [36](vscode-notebook-cell:?execution_count=27&line=36) assert np.all(output_collapse_axis1["sum"] == expected_sum_result_axis1)

AssertionError:
```

Did I misunderstand the method?

**Comment by fsenf** [comment]
Just forget about my last comment - I was mixing up column and row order ... sorry!

**Comment by w-k-jones** [comment]
@fsenf sorry, I should have commented what I was doing a bit better to make things clearer!

The `collapse_axis` keyword uses the same axis numbering convention as numpy, so in your test case `collapse_axis=1` means it is collapsing the rows of the input array. As shown by `test_labels.max(axis=1) == array([0, 1, 1, 2, 0])`, this means that the 2nd feature is collapsed into only the 4th row, and so the expected result should be `[2.25, 1.0]`. If you change to `collapse_axis=0` then you should get the result `[2.25, 0.75]`

File: tobac/__init__.py
Comment (w-k-jones) [reply tobac/__init__.py]: I've split up the `analysis.analysis` module into `cell_analysis` and `feature_analysis` to separate them a bit better, but I'm still not sure what a better name for them is. I didn't want to use `statistics` to avoid confusion with the `bulk_statistics` module! I've added `__all__` statements to both and added imports to the `__init__.py` file to ensure back-compatibility

**Comment by w-k-jones** [comment]
Yes, it's due to a bug in older versions of `xarray` that both @freemansw1 and I have run into which causes conversion of integer arrays from `iris`->`xarray`->`iris` to fail. I'll see if there's an easy way to handle it

**Comment by w-k-jones** [comment]
Right, I think I've got to the bottom of this. As @freemansw1 found, this is due to a bug in `xarray` that has since been partially fixed (since `v2023.06`). It effectively has two parts: 1. When converting from `xarray` to `iris` the data is always converted to a masked array. 2. When converting from `iris` to `xarray` with a masked array of data `np.nan` is always used as the fill value for floating and integer data, regardless of whether the data has a fill_value assigned. The "fix" in `xarray` is due to the way in which `np.nan` is handled when casting to an integer array by `numpy`:

If you cast `np.nan` directly to integer it will raise an exception:
```
np.array(np.nan, np.int32)
ValueError                                Traceback (most recent call last)
Cell In[29], line 1
----> 1 np.array(np.nan, np.int32)

ValueError: cannot convert float NaN to integer
```

But if you first cast `np.nan` to the corresponding _floating_ dtype (which occurs in `xarray` post `v2023.06`) it runs with a warning:
```
np.array(np.float32(np.nan), np.int32)
RuntimeWarning: invalid value encountered in cast
  np.array(np.float32(np.nan), np.int32)
array(-2147483648, dtype=int32)
```

I can see two solutions to this. The simple one (and probably best long term) is to require `xarray>=2023.06` in the requirements. The second is to check when converting from iris to xarray whether the data is a masked array and fill correctly before converting, but this is slightly more complex.

**Comment by fsenf** [comment]
@w-k-jones : Unfortunately, the notebook is still failing to run through after your fix :-(

**Comment by w-k-jones** [comment]
I haven’t addressed the issue with the notebook yet, still deciding what action to take. Pinning the xarray requirements is the easiest but I generally try to avoid doing that

**Comment by fsenf** [comment]
after #411 is done, this PR could be the next one. @w-k-jones : Any idea for a quickfix for the notebook bug?

**Comment by w-k-jones** [comment]
I have added a fix to the issues with iris->xarray conversion (see `convert_cube_to_dataarray` in `utils.decorators`) which handles the situation in which a masked, integer cube would cause an error when converting to xarray. The notebooks now seem to all run correctly, even with older versions of xarray. I have also included updates from RC_v1.5.x and the fixes in #416 , so will wait for that to be merged before this.

@fsenf would you bee able to review?

**Comment by fsenf** [comment]
Yes, great job, @w-k-jones ! CI looks good! We will allocate more time for review tomorrow.

**Review by fsenf — APPROVED** [review]
Dear Will,

really great work! Esp. the expansion of testing capabilities (~1000 new lines) is exceptional!

I marked a small inconsistency, but I am also happy if you just proceed and merge you devs back to the release candidate.

Cheers, Fabian.

File: tobac/tests/test_analysis_spatial.py
Comment (w-k-jones) [reply tobac/tests/test_analysis_spatial.py:365]: In this case it's meant to be inconsistent, as I was trying to test the error condition for when a dataset has 1D latitude but 2D longitude (i.e. inconsistent dimensionality on lat/lon). I will update the dimension naming to make it a bit clearer however

**Comment by w-k-jones** [comment]
Thanks @fsenf , I will update the case you highlighted to make my intentions clearer. Will wait until after #416 to merge

**Comment by freemansw1** [comment]
> Thanks @fsenf , I will update the case you highlighted to make my intentions clearer. Will wait until after #416 to merge

I've just merged #416

**Pull request merged**
